### PR TITLE
Storage adapters

### PR DESCRIPTION
**Pull request by pachico**
Please help me validate the idea.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -32,7 +32,7 @@
     ],
     "require": {
         "php": "~5.5|~7.0",
-        "webmozart/assert": "1.0.2"
+        "webmozart/assert": "1.0.*"
```
Comment (pachico) [inline composer.json:35]: Reason for this is that new Composer fails if version definition is too strict.

File: src/Storage/Adapter/Runtime.php [hunk 2]
```diff
@@ -26,16 +24,16 @@ class Runtime implements StorageInterface
 
     /**
      * Initializes a new instance of this class.
+     *
+     * @param array $data
      */
-    public function __construct()
+    public function __construct(array $data = [])
```
Comment (pachico) [inline src/Storage/Adapter/Runtime.php:29]: I added this so we can easily use it for tests where you need certain data to be already stored.

File: example/01-bootstrap.php [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+$test2->addVariant(new SimpleVariant('_variant2'));
+
+// Add the tests to the Engine
+$engine->addTest($test);
+$engine->addTest($test2);
+
+// Pseudo: if($user->isAdmin)
+// If the user is admin, he should not participate at the test
+// $manager->participate('foo_test', null);
+// Pseudo: if($app->inDevelopment() and $GET['phpab']['foo_test])
```
Comment (waltertamboer) [inline example/01-bootstrap.php]: $_GET :-)

File: example/01-bootstrap.php [hunk 4]
```diff
@@ -0,0 +1,76 @@
+<?php
+
```
Comment (waltertamboer) [inline example/01-bootstrap.php]: Please remove this blank line. Docblocks of the file should be placed directly after the PHP tag (all other files have that as well).

File: src/Storage/Adapter/AdapterInterface.php [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+namespace PhpAb\Storage\Adapter;
+
+/**
+ * Used by StorageInterface to store user participations state
+ *
+ * @package PhpAb
+ */
+interface AdapterInterface
+{
+
```
Comment (waltertamboer) [inline src/Storage/Adapter/AdapterInterface.php]: Unneeded blank line.

File: src/Storage/Adapter/AdapterInterface.php [hunk 6]
```diff
@@ -0,0 +1,71 @@
+<?php
+
```
Comment (waltertamboer) [inline src/Storage/Adapter/AdapterInterface.php]: Unneeded blank line.

File: src/Storage/Adapter/Cookie.php [hunk 7]
```diff
@@ -1,4 +1,5 @@
 <?php
+
```
Comment (waltertamboer) [inline src/Storage/Adapter/Cookie.php]: Not needed

File: src/Storage/Adapter/Cookie.php [hunk 8]
```diff
@@ -66,76 +65,64 @@ public function __construct($cookieName, $ttl = 157766400)
...
     public function has($identifier)
     {
         Assert::string($identifier, 'Test identifier is invalid.');
         Assert::notEmpty($identifier, 'Test identifier is invalid.');
-        
+
         $this->parseExistingCookie();
 
-        return isset($this->cookieValues[$identifier]);
+        return isset($this->data[$identifier]);
```
Comment (waltertamboer) [inline src/Storage/Adapter/Cookie.php]: I think we should use `array_key_exists`, isset fails if the value is '0' for example.

File: src/Storage/Adapter/Runtime.php (hunk 2 above)
Comment (waltertamboer) [inline src/Storage/Adapter/Runtime.php:29]: :+1:

File: src/Storage/Storage.php [hunk 9]
```diff
@@ -0,0 +1,61 @@
+<?php
+
```
Comment (waltertamboer) [inline src/Storage/Storage.php]: Not needed.

File: src/Storage/Storage.php [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+use PhpAb\Storage\Adapter\AdapterInterface;
+
+/**
+ * {@inheritDoc}
+ */
+class Storage implements StorageInterface
+{
+    /**
+     * @var Adapter\AdapterInterface
```
Comment (waltertamboer) [inline src/Storage/Storage.php]: AdapterInterface is enough, you have a use statement.

File: src/Storage/Storage.php [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+
+    /**
+     * {@inheritDoc}
+     */
+    public function set($identifier, $participation)
+    {
+        $this->adapter->set($identifier, $participation);
+    }
+
+    public function clear()
```
Comment (waltertamboer) [inline src/Storage/Storage.php:59]: Misses a docblock.

File: src/Storage/StorageInterface.php [hunk 12]
```diff
@@ -1,4 +1,5 @@
 <?php
+
```
Comment (waltertamboer) [inline src/Storage/StorageInterface.php]: Unneeded.

File: src/Storage/StorageInterface.php [hunk 13]
```diff
@@ -16,6 +17,7 @@
  */
 interface StorageInterface
 {
+
```
Comment (waltertamboer) [inline src/Storage/StorageInterface.php]: Also not needed.

**Comment by waltertamboer** [comment]
It looks good, I think this will be a good change :+1:

File: src/Storage/Adapter/Cookie.php (hunk 8 above)
Comment (pachico) [inline src/Storage/Adapter/Cookie.php]: Well seen!

**Pull request merged**